runtime/io.c

```diff
@@ -139,34 +150,32 @@ CAMLexport void caml_close_channel(struct channel *channel)
 
 CAMLexport file_offset caml_channel_size(struct channel *channel)
 {
-  file_offset offset;
-  file_offset end;
+  file_offset here, end;
   int fd;
-  check_pending(channel);
 
+  check_pending(channel);
   /* We extract data from [channel] before dropping the OCaml lock, in case
      someone else touches the block. */
   fd = channel->fd;
-  offset = channel->offset;
+  here = channel->flags & CHANNEL_TEXT_MODE ? -1 : channel->offset;
```

Comment:
FTR, one could probably use `GetFileSizeEx` (+ `_get_osfhandle`) under Win32 to avoid the lseek dance, but I see no strong reason to switch to it (perhaps performance is a bit better, but we probably don't care).